src/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,278 @@
+    options: ActionListenerOptions<any, any, any>
+  },
+  'actionListenerMiddleware/add'
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D>,
+    options?: ActionListenerOptions<ReturnType<C>, S, D>
+  ): AddListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D>,
+    options?: ActionListenerOptions<AnyAction, S, D>
+  ): AddListenerAction<AnyAction, S, D>
+}
+
+interface RemoveListenerAction<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>
+> {
+  type: 'actionListenerMiddleware/remove'
+  payload: {
+    type: string
+    listener: ActionListener<A, S, D>
+  }
+}
+
+/**
+ * @alpha
+ */
+export const removeListenerAction = createAction(
+  'actionListenerMiddleware/remove',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type,
+        listener
+      }
+    }
+  }
+) as BaseActionCreator<
+  { type: string; listener: ActionListener<any, any, any> },
+  'actionListenerMiddleware/remove'
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D>
+  ): RemoveListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D>
+  ): RemoveListenerAction<AnyAction, S, D>
+}
+
+/**
+ * @alpha
+ */
+export function createActionListenerMiddleware<
+  S,
+  D extends Dispatch<AnyAction> = Dispatch
+>() {
+  type ListenerEntry = ActionListenerOptions<any, S, D> & {
+    listener: ActionListener<any, S, D>
+  }
+
+  const listenerMap: Record<string, Set<ListenerEntry> | undefined> = {}
+  const middleware: Middleware<
+    {
+      (action: Action<'actionListenerMiddleware/add'>): Unsubscribe
+    },
+    S,
+    D
+  > = api => next => action => {
+    if (addListenerAction.match(action)) {
+      const unsubscribe = addListener(
+        action.payload.type,
+        action.payload.listener,
+        action.payload.options
+      )
+
+      return unsubscribe
+    }
+    if (removeListenerAction.match(action)) {
+      removeListener(action.payload.type, action.payload.listener)
+
+      return
+    }
+
+    const listeners = listenerMap[action.type]
+    if (listeners) {
+      for (const entry of listeners) {
+        if (!entry.condition || entry.condition(action, api.getState)) {
+          entry.listener(action, api)
+          if (entry.once) {
+            listeners.delete(entry)
+          }
+          if (entry.preventPropagation) {
+            return action
+          }
+        }
+      }
+    }
+    return next(action)
```

Comment:
The `next(action)` call occurs after calling the actions with `entry.listener(action, api)`.  This means that you can't have a listener that uses the updated state, right?  Is this intentional?  I found https://github.com/reduxjs/redux-toolkit/issues/237#issuecomment-562287869 that says the reducers should run first.  The logic for this block might get tricky with checking conditions and preventPropagation, but might still be doable.

Reply:
Phew. That's a good point.
Ideally, we'd like to allow for both behaviours.

I could imagine adding an option `{ when: 'before' | 'after' }`, defaulting it to `before` and issuing an Error from the `stopPropagation` method when called in combination with `after`.
What do you think?

Reply:
Ouch. This one hurts quite a bit in the typings. Following through with it nevertheless ^^

Reply:
Okay, that's in. Comments? :)

Reply:
If when is after, should the entry.condition be given the previous state or the next state?
You are returning from within the for loop which means not all listeners will be processed.

This is a bit verbose, and totally untested, but how about something like:
```TypeScript
if (listeners) {
  // process 'before' listeners
  for (const entry of listeners) {
    if (entry.when == 'after') {
      continue;
    }

    if (!entry.condition || entry.condition(action, api.getState)) {
      if (entry.once) {
        listeners.delete(entry)
      }

      let stoppedPropagation = false
      entry.listener(action, {
        ...api,
        stopPropagation() {
          stoppedPropagation = true
        }
      })
      if (stoppedPropagation) {
        return action
      }
    }
  }

  const result = next(action)

  // process 'after' listeners
  for (const entry of listeners) {
    if (entry.when != 'after') {
      continue;
    }

    if (!entry.condition || entry.condition(action, api.getState)) {
      if (entry.once) {
        listeners.delete(entry)
      }

      entry.listener(action, {
        ...api,
        stopPropagation: () => {
          throw new Error(
          'stopPropagation can only be called by action listeners with the `when` option set to "before"'
          )
        }
      })
    }
  }
  return result
}
return next(action)
```

I think this leaves the door open for having a `when: both` option, but I don't know why someone would want that.

Reply:
I think if someone wanted to do `when: both`, they could just add a `before` listener and an `after` listener. We really shouldn't add any extra logic for something like that.

But you are very right about the condition. That should definitely run after for the `after` case - good catch!

Reply:
Okay, I pushed a fix and added another two tests for that. Could you please take another look? :)

Reply:
That looks good.  I'm still unsure about which default (before or after) is better.  For people coming from Saga, having after be the default might be less confusing.  That said, if there's a good case for running before the reducers then it's probably better to default to before.  I'll be curious to hear more thoughts on this.

Reply:
Personally, I'd like to have the ability to `stopPropagation` by default.  
I believe in sagas it's only the case because those are async AF and usually a million ticks have happened yielding 80 times before anything meaningful happens. (Doesn't sound like it, but I still really like saga ;) )

@markerikson do you have an opinion on this?

Reply:
Both sagas and observables explicitly let the action hit the reducers first:

https://github.com/redux-saga/redux-saga/blob/a8c968b77243f59fbe672ca56691761664e060bf/packages/core/src/internal/middleware.js#L27

https://github.com/redux-observable/redux-observable/blob/dfef883fdba9f669ea88139bdb7abd2d026ff7a8/src/createEpicMiddleware.ts#L88-L91

I'd say that's the preferable default.

Reply:
Just a thought, could the listener behave a bit like `useEffect` and allow you to return a function to run after the state update? something like:

```js
function myListener(action, { getState() }) {
  let previousValue = getState().valueWeCareAbout
  return () => {
    let  newValue =  getState().valueWeCareAbout
    if (previousValue === 'Pending' && newValue === 'Failed') {
      logger.error('Failed to submit')
    }
  }
}
```

Then you could have before listener

```js
const beforeListener = () => {}
```

and after listener

```js
const afterListener = () => () => {}
```

or a before and after listener

```js
const bothListener = () => {
  return () => {}
}
```

Reply:
That also feels kinda complex, and the similarity in syntax with `useEffect` would actually seem like a reason to _not_ do it that way because they're two very different things.

Reply:
I guess I don't see them as _very_ different in terms of wanting to perform logic around state changing (dependency array for `useEffect`, redux state for this).  I could also argue that this primary use case for this middleware is to use effects in response to actions.

There are also some similarities in the api with exisiting redux concepts with thunks and middleware.  It could be considered a middleware-lite api so converting your listener to a custom middleware when the requirements grow is eased a bit.